src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala

```diff
@@ -406,9 +407,10 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
 
       // must do this before removing guards from cases and collapsing (scala/bug#6011, scala/bug#6048)
       def unreachableCase(cases: List[CaseDef]): Option[CaseDef] = {
+        @tailrec
         def loop(cases: List[CaseDef]): Option[CaseDef] = cases match {
           case head :: next :: _ if isDefault(head)                                    => Some(next) // subsumed by the next case, but faster
-          case head :: rest if !isGuardedCase(head) || head.guard.tpe =:= ConstantTrue => rest find caseImplies(head) orElse loop(rest)
+          case head :: rest if !isGuardedCase(head) || head.guard.tpe =:= ConstantTrue => rest find caseImplies(head) match { case s @ Some(_) => s case None => loop(rest) }
```

Comment:
Hand inlined `orElse` to enable tail call elimination.

Reply:
I'm especially impressed by the hardened resistance to semicolon.